Tools/cases_generator/stacking.py

```diff
@@ -0,0 +1,388 @@
+from abc import abstractmethod
+import dataclasses
+import typing
+
+from formatting import (
+    Formatter,
+    UNUSED,
+    maybe_parenthesize,
+    parenthesize_cond,
+)
+from instructions import (
+    ActiveCacheEffect,
+    Instruction,
+    MacroInstruction,
+    Component,
+    Tiers,
+    TIER_ONE,
+)
+from parsing import StackEffect, CacheEffect, Family
+
+
+@dataclasses.dataclass
+class StackOffset:
```

Comment:
Could this be represented instead as an index from the TOS, then a capture of the "stack"?

E.g. for a `PEEK(1)` it would be `index=-2, stack=[item1, item2, item3, item4]`.

Reply:
Hm... That's closer to the old way of doing this, where the stack was represented (implicitly) by variables `_tmp_1`, `_tmp_2`, etc., and the stack offset of an effect was represented as being mapped to one of those variables (using the old `input_mapping` and `output_mapping` members of `Component`). I ran into some problems there when an effect is conditional, and managed to hack that in (but only for the output effects of the last component). But with array effects for guard instructions like we will need for CALL guards I just couldn't hack it any more, and instead I came up with this abstraction, which can handle conditional and array effects anywhere in a macro.

The only thing this cannot handle is a situation where a value is temporarily pushed onto the stack, stays there for the next uop, and then is popped later:
```
op(A, (-- temp)) { ... }  // stack: [] -> [temp]
op(B, (--)) { ... }       // stack: [temp] -> [temp]
op(C, (temp --)) { ... }  // stack: [temp] -> []
macro(M1) = A + C;
macro(M2) = A + B + C;
```
Here both `M1` and `M2` have a _net_ stack effect of 0, `n_popped` is 0, and `n_pushed` is 0, and we cannot express using `n_popped` and `n_pushed` that this uses one temporary stack item. For `M1` that's not a problem, because the algorithm translates the push in `A` and the pop in `C` into a copy, which doesn't require stack space. (Also, the copy disappears because the variable name is the same.) But for `M2` the push and pop are not adjacent so they are not optimized away like that.

I *could* improve the algorithm to recognize this situation and use a copy for M2, but it's more complicated and it's unlikely that we'll need this. (Most likely in a real case there would be a result pushed onto the stack at the end, so the problem of phantom stack space wouldn't occur.) I *ought to* at least detect it and warn, but I'd rather do that in a future PR, since this one is complex enough as it is.

Note: the generated code for `M2` clearly shows the problem:
```
        TARGET(M2) {
            PyObject *temp;
            // A
            {
                 ...
            }
            stack_pointer[0] = temp;
            // B
            {
                 ...
            }
            // C
            temp = stack_pointer[0];
            {
                 ...
            }
            DISPATCH();
        }
```
Note that `stack_pointer[0]` is an invalid stack item, pointing just *above* the current stack top. (The actual top is `stack_pointer[-1]`.)